cranelift/codegen/src/machinst/reg.rs

```diff
@@ -0,0 +1,432 @@
+//! Definitions for registers, operands, etc. Provides a thin
+//! interface over the register allocator so that we can more easily
+//! swap it out or shim it when necessary.
+
+use crate::machinst::MachInst;
+use alloc::{string::String, vec::Vec};
+use core::{fmt::Debug, hash::Hash};
+use regalloc2::{Allocation, Operand, PReg, VReg};
+use smallvec::{smallvec, SmallVec};
+
+/// The first 128 vregs (64 int, 64 float/vec) are "pinned" to
+/// physical registeres: this means that they are always constrained
+/// to the corresponding register at all use/mod/def sites.
+///
+/// Arbitrary vregs can also be constrained to physical registers at
+/// particular use/def/mod sites, and this is preferable; but pinned
+/// vregs allow us to migrate code that has been written using
+/// RealRegs directly.
+pub const PINNED_VREGS: usize = 128;
+
+/// Convert a `VReg` to its pinned `PReg`, if any.
+pub fn pinned_vreg_to_preg(vreg: VReg) -> Option<PReg> {
+    if vreg.vreg() < PINNED_VREGS {
+        Some(PReg::from_index(vreg.vreg()))
+    } else {
+        None
+    }
+}
+
+/// A register named in an instruction. This register can be either a
+/// virtual register or a fixed physical register. It does not have
+/// any constraints applied to it: those can be added later in
+/// `MachInst::get_operands()` when the `Reg`s are converted to
+/// `Operand`s.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Reg(VReg);
+
+impl Reg {
+    /// Extract the underlying `regalloc2::VReg`. Note that physical
+    /// registers also map to particular (special) VRegs, so this
+    /// method can be used either on virtual or physical `Reg`s.
+    pub fn to_vreg(self) -> VReg {
+        self.0
+    }
+
+    /// Create a `Reg` from an underlying `regalloc2::VReg`.
+    pub fn from_vreg(vreg: VReg) -> Self {
+        Self(vreg)
+    }
+
+    /// Get the physical register (`RealReg`), if this register is
+    /// one.
+    pub fn to_real_reg(self) -> Option<RealReg> {
+        if self.0.vreg() < PINNED_VREGS {
+            Some(RealReg(self.0))
+        } else {
+            None
+        }
```

Comment:
Rather than open coding the `< PINNED_VREGS` logic here and elsewhere, can we define a helper and make this something like

```rust
if is_pinned(self.0.vreg()) {
    Some(RealReg(self.0))
} else {
    None
}
```

and similar below. This way if we change how we encode pinned registers, we don't need to update a bunch of different open-coded places.

Reply:
Good idea; I've refactored in this direction. This particular helper (`pinned_vreg_to_preg`) is the canonical helper that defines the index split, along with `first_user_vreg_index()` below (index needed for vreg index allocation during lowering, but exposing it with this name only makes its intended use more clear).